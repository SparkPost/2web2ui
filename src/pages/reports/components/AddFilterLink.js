import React, { useState } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import qs from 'qs';

import { addFilters } from 'src/actions/reportOptions';
import { PageLink } from 'src/components/links';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import {
  selectReportSearchOptions,
  selectSummaryChartSearchOptions,
} from 'src/selectors/reportSearchOptions';

export const AddFilterLink = ({
  //From parent
  newFilter,
  reportType,
  content,
  //From redux
  currentSearchOptions,
  addFilters,
}) => {
  const [isNewTabKeyPressed, setNewTabKeyPressed] = useState(false);

  const currentFilters = currentSearchOptions.filters || [];

  /**
   * Needs to first check if cmd/ctrl key is pressed when mouse is released
   * Then it's saved for use in the onClick handler. We need both because
   * onClick does not have the correct values for metaKey and ctrlKey
   **/
  const handleMouseUp = e => {
    setNewTabKeyPressed(e.metaKey || e.ctrlKey);
  };

  const handleClick = () => {
    if (!isNewTabKeyPressed) {
      addFilters([newFilter]);
    }
  };

  const mergedFilters = _.uniqWith(
    [...currentFilters, stringifyTypeaheadfilter(newFilter)],
    _.isEqual,
  );
  const newSearchOptions = { ...currentSearchOptions, filters: mergedFilters };
  const linkParams = qs.stringify(newSearchOptions, { arrayFormat: 'repeat' });
  const fullLink = `/reports/${reportType}/?${linkParams}`;

  return (
    <PageLink onMouseUp={handleMouseUp} onClick={handleClick} to={fullLink} replace>
      {content}
    </PageLink>
  );
};

const mapStateToProps = (state, props) => ({
  currentSearchOptions:
    props.reportType === 'summary'
      ? selectSummaryChartSearchOptions(state)
      : selectReportSearchOptions(state),
});

export default connect(mapStateToProps, { addFilters })(AddFilterLink);
