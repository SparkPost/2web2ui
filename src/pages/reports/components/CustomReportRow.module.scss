@import '~@sparkpost/matchbox/src/styles/config.scss';

.Row {
  display: flex;
  border-bottom: 1px solid color(gray, 9);

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    cursor: pointer;
  }
}

.Delete {
  flex: 0 0;
  align-self: flex-end;
  padding: spacing();
}

.DeleteButton {
  color: color(gray, 4);

  &:hover {
    color: color(red)
  }
}

.Name {
  flex: 1 0;
  padding: spacing();
  line-height: line-height(600);
  font-size: font-size(500);

  color: color(blue, dark);

  &:hover {
    text-decoration: underline;
  }
}
