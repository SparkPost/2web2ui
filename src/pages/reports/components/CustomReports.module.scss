@import '~@sparkpost/matchbox/src/styles/config.scss';

.NameWrapper {
  input {
    font-size: font-size(500);
  }
}

.Wrap {
  display: flex;
  align-items: baseline;

  & > *:not(:first-child) {
    margin-left: spacing();
  }
}

.ReportsButton {
  flex: 1 0;
  text-align: right;
}

.SaveIcon {
  margin-top: -0.2em;
  margin-right: 0.5ch;
}

.Limited {
  font-size: font-size(200);
  line-height: line-height(500);
  color: color(gray, 5);
}
