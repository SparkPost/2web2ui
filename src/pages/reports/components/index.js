export { default as ReportOptions } from './ReportOptions';
export { default as MetricCard } from './MetricCard';
export { default as MetricsSummary } from './MetricsSummary';
export { default as ShareModal } from './ShareModal';
