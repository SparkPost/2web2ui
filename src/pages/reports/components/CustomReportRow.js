import React from 'react';
import { Button } from 'src/components/matchbox';
import styles from './CustomReportRow.module.scss';

function ReportRow(props) {
  return (
    <div onClick={props.onLoad} className={styles.Row}>
      <div className={styles.Name}>{props.name}</div>
      <div className={styles.Delete}>
        <Button
          className={styles.DeleteButton}
          disabled={props.deleting}
          flat
          size="small"
          onClick={e => {
            e.stopPropagation();
            props.onDelete(props.name);
          }}
        >
          Delete
        </Button>
      </div>
    </div>
  );
}

export default ReportRow;
