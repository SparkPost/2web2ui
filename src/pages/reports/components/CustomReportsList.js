import React from 'react';
import { withRouter } from 'react-router-dom';
import { Button, Panel, Modal } from 'src/components/matchbox';
import ReportRow from './CustomReportRow';

export function CustomReportsList({ deleting, handleLoad, onDelete, reports }) {
  const [open, setOpen] = React.useState(false);

  const reportList = React.useMemo(() => {
    if (!reports.length) {
      return <Panel.LEGACY.Section>You do not have any saved reports.</Panel.LEGACY.Section>;
    }

    return reports.map((report, i) => {
      function onLoad() {
        setOpen(false);
        handleLoad(report);
      }

      return (
        <ReportRow
          key={`${report.name}-${i}`}
          {...report}
          onLoad={onLoad}
          onDelete={onDelete}
          deleting={deleting}
        />
      );
    });
  }, [deleting, handleLoad, onDelete, reports]);

  return (
    <>
      <Button flat color="orange" size="small" onClick={() => setOpen(!open)}>
        Saved Reports
      </Button>
      <Modal.LEGACY open={open} onClose={() => setOpen(false)} showCloseButton>
        <Panel.LEGACY title="Saved Reports">{reportList}</Panel.LEGACY>
      </Modal.LEGACY>
    </>
  );
}

export default withRouter(CustomReportsList);
