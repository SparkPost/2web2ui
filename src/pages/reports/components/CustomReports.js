import React from 'react';
import { connect } from 'react-redux';
import { addFilters, clearFilters, refreshReportOptions } from 'src/actions/reportOptions';
import { selectCustomReports } from 'src/selectors/customReports';
import { Save } from '@sparkpost/matchbox-icons';
import { saveReport, deleteReport } from 'src/actions/customReports';
import { withRouter } from 'react-router-dom';
import { Button, Panel, TextField } from 'src/components/matchbox';
import CustomReportsList from './CustomReportsList';
import { parseSearch } from 'src/helpers/reports';
import qs from 'query-string';
import _ from 'lodash';

import styles from './CustomReports.module.scss';

const REPORT_LIMIT = 10;

export function CustomReports(props) {
  const {
    addFilters,
    deleteReport,
    clearFilters,
    searchOptions,
    loading,
    refreshReportOptions,
    reports,
    saveReport,
  } = props;

  const [name, setName] = React.useState('');

  function handleNameChange(e) {
    const value = e.target.value;

    // Limit string length
    if (value.length < 24) {
      setName(e.target.value);
    }
  }

  function handleSave() {
    let toSave = searchOptions;

    if (searchOptions.range !== 'custom') {
      toSave = _.omit(toSave, ['to', 'from']);
    }

    saveReport({
      name: name.trim(),
      url: qs.stringify(toSave),
    });
  }

  function handleLoad(report) {
    const { options, filters = [] } = parseSearch(report.url);
    clearFilters();
    addFilters(filters);
    refreshReportOptions(options);
    setName(report.name);
  }

  const isLimited = reports.length >= REPORT_LIMIT;
  const canOverwrite = _.find(reports, ['name', name.trim()]);
  const saveDisabled = name === '' || loading;

  return (
    <Panel.LEGACY.Section>
      <div className={styles.Wrap}>
        <div className={styles.NameWrapper}>
          <TextField
            id="custom-report-name"
            label="Report Name"
            labelHidden
            placeholder="Untitled Report"
            onChange={handleNameChange}
            value={name}
          />
        </div>

        {!isLimited || (canOverwrite && isLimited) ? (
          <Button disabled={saveDisabled} onClick={handleSave} flat color="orange">
            <Save className={styles.SaveIcon} />
            {loading ? 'Saving...' : 'Save'}
          </Button>
        ) : null}

        {isLimited && (
          <div className={styles.Limited}>You are limited to {REPORT_LIMIT} saved reports</div>
        )}

        <div className={styles.ReportsButton}>
          <CustomReportsList
            deleting={loading}
            handleLoad={handleLoad}
            reports={reports}
            onDelete={deleteReport}
          />
        </div>
      </div>
    </Panel.LEGACY.Section>
  );
}

const mapStateToProps = state => ({
  reports: selectCustomReports(state),
  loading: state.currentUser.userOptionsPending,
});

const mapDispatchToProps = {
  addFilters,
  clearFilters,
  refreshReportOptions,
  saveReport,
  deleteReport,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CustomReports));
