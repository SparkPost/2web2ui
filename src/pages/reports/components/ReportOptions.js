import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import {
  addFilters,
  removeFilter,
  refreshReportOptions,
  refreshTypeaheadCache,
  initTypeaheadCache,
} from 'src/actions/reportOptions';
import ShareModal from './ShareModal';
import PrecisionSelector from './PrecisionSelector';
import { parseSearch } from 'src/helpers/reports';
import { isForcedUTCRollupPrecision } from 'src/helpers/metrics';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import Typeahead from './Typeahead';
import { Grid, Panel, Select, Tag, Tooltip } from 'src/components/matchbox';
import DatePicker from 'src/components/datePicker/DatePicker';
import typeaheadCacheSelector from 'src/selectors/reportFilterTypeaheadCache';
import { TimezoneTypeahead } from 'src/components/typeahead/TimezoneTypeahead';
import CustomReports from './CustomReports';
import _ from 'lodash';
import config from 'src/config';
import OGStyles from './ReportOptions.module.scss';
import hibanaStyles from './ReportOptionsHibana.module.scss';

const { metricsRollupPrecisionMap } = config;
const RELATIVE_DATE_OPTIONS = ['hour', 'day', '7days', '30days', '90days', 'custom'];
const PRECISION_OPTIONS = metricsRollupPrecisionMap.map(({ value }) => ({
  value,
  label: _.startCase(_.words(value).join(' ')),
}));

export class ReportOptionsClassComponent extends Component {
  state = {
    shownPrecision: '',
  };

  componentDidMount() {
    const { options, filters = [] } = parseSearch(this.props.location.search);
    this.props.addFilters(filters);
    this.props.refreshReportOptions(options);

    // initial typeahead cache load
    this.props.initTypeaheadCache();
  }

  updateShownPrecision = shownPrecision => {
    this.setState({ shownPrecision });
  };

  renderActiveFilters = () => {
    const { reportOptions, styles } = this.props;

    return reportOptions.filters.length ? (
      <Panel.LEGACY.Section>
        <small>Filters:</small>
        {reportOptions.filters.map((item, index) => (
          <Tag
            key={index}
            onRemove={() => this.handleFilterRemove(index)}
            className={styles.TagWrapper}
          >
            {item.type}: {item.value}
          </Tag>
        ))}
      </Panel.LEGACY.Section>
    ) : null;
  };

  handleFilterRemove = index => {
    this.props.removeFilter(index);
  };

  handleTypeaheadSelect = item => {
    this.props.addFilters([item]);
  };

  handleTimezoneSelect = timezone => {
    const { refreshReportOptions } = this.props;
    refreshReportOptions({ timezone: timezone.value });
  };

  getPanelContent = () => {
    const {
      typeaheadCache,
      reportOptions,
      reportLoading,
      refreshReportOptions,
      searchOptions,
      styles,
    } = this.props;

    const isForcedUTC =
      reportOptions.precision && isForcedUTCRollupPrecision(reportOptions.precision);

    const isShownForcedUTC =
      this.state.shownPrecision && isForcedUTCRollupPrecision(this.state.shownPrecision);

    const timezoneDisabled = reportLoading || (isForcedUTC && this.state.shownPrecision === '');

    const timezoneTypeahead = (
      <TimezoneTypeahead
        initialValue={reportOptions.timezone}
        onChange={this.handleTimezoneSelect}
        isForcedUTC={isForcedUTC}
        disabledAndUTCOnly={!!isShownForcedUTC}
        disabled={timezoneDisabled}
      />
    );

    return (
      <>
        <Panel.LEGACY.Section>
          <Grid>
            <Grid.Column xs={9} md={6}>
              <Typeahead
                reportOptions={reportOptions}
                placeholder="Filter by domain, campaign, etc"
                onSelect={this.handleTypeaheadSelect}
                items={typeaheadCache}
                selected={reportOptions.filters}
                disabled={reportLoading}
              />
            </Grid.Column>
            <Grid.Column xs={3} md={2} mdOffset={4}>
              <ShareModal disabled={reportLoading} searchOptions={searchOptions} />
            </Grid.Column>
          </Grid>
        </Panel.LEGACY.Section>
        <Panel.LEGACY.Section>
          <Grid>
            <Grid.Column xs={12} md={6}>
              <div className={styles.FieldWrapperMetricsRollup}>
                <DatePicker
                  {...reportOptions}
                  label="Date Range"
                  relativeDateOptions={RELATIVE_DATE_OPTIONS}
                  disabled={reportLoading}
                  onChange={refreshReportOptions}
                  roundToPrecision={true}
                  selectPrecision={true}
                  updateShownPrecision={this.updateShownPrecision}
                />
              </div>
            </Grid.Column>
            <Grid.Column xs={6} md={4}>
              <div className={styles.TimezoneTooltipWrapper}>
                <Tooltip
                  disabled={!isShownForcedUTC && !timezoneDisabled}
                  content="Day, week, and month precision only support UTC."
                >
                  {timezoneTypeahead}
                </Tooltip>
              </div>
            </Grid.Column>
            <Grid.Column xs={6} md={2}>
              {//We will show a fake selector that shows the temporary precision when the user
              //is selecting dates using the datepicker but has not confirmed the selection
              !this.state.shownPrecision ? (
                <PrecisionSelector
                  from={reportOptions.from}
                  to={reportOptions.to}
                  selectedPrecision={reportOptions.precision}
                  changeTime={refreshReportOptions}
                  disabled={reportLoading}
                />
              ) : (
                <Select
                  id="precision-select"
                  label="Precision"
                  options={PRECISION_OPTIONS}
                  value={this.state.shownPrecision}
                  disabled={reportLoading}
                  readOnly
                />
              )}
            </Grid.Column>
          </Grid>
        </Panel.LEGACY.Section>
      </>
    );
  };

  render() {
    const { customReportsEnabled, searchOptions } = this.props;

    return (
      <div data-id="report-options">
        <Panel.LEGACY>
          {customReportsEnabled && <CustomReports searchOptions={searchOptions} />}
          {this.getPanelContent()}
          {this.renderActiveFilters()}
        </Panel.LEGACY>
      </div>
    );
  }
}

// TODO: Wrapping component just until the OG theme is removed to pass in styles - won't be needed in the future.
function ReportOptions(props) {
  const styles = useHibanaOverride(OGStyles, hibanaStyles);

  return <ReportOptionsClassComponent styles={styles} {...props} />;
}

const mapStateToProps = state => ({
  reportOptions: state.reportOptions,
  typeaheadCache: typeaheadCacheSelector(state),
});

const mapDispatchToProps = {
  addFilters,
  removeFilter,
  refreshReportOptions,
  initTypeaheadCache,
  refreshTypeaheadCache,
};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ReportOptions));
