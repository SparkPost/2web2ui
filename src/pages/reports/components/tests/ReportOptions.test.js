import React from 'react';
import { shallow } from 'enzyme';
import { ReportOptionsClassComponent as ReportOptions } from '../ReportOptions';
import Typeahead from '../Typeahead';
import { Tag } from 'src/components/matchbox';
import * as reportHelpers from 'src/helpers/reports';
import styles from '../ReportOptions.module.scss';

jest.mock('src/helpers/date');
jest.mock('src/helpers/reports');

describe('Component: Report Options', () => {
  let wrapper;
  let testProps;
  let filters;
  let options;
  let testQuery;

  beforeEach(() => {
    const testDate = new Date('2018-02-15T12:00:00Z');
    options = {};
    filters = [];
    testQuery = {};
    reportHelpers.getReportSearchOptions = jest.fn(() => testQuery);
    reportHelpers.parseSearch = jest.fn(() => ({ options, filters }));
    testProps = {
      addFilters: jest.fn(),
      removeFilter: jest.fn(),
      typeaheadCache: [],
      reportOptions: {
        from: testDate,
        to: testDate,
        relativeRange: 'day',
        filters,
      },
      location: {
        search: '?some=test',
        pathname: 'my-pathname',
      },
      history: {
        replace: jest.fn(),
      },
      initTypeaheadCache: jest.fn(),
      refreshReportOptions: jest.fn(),
      refreshTypeaheadCache: jest.fn(),
      styles,
    };
    wrapper = shallow(<ReportOptions {...testProps} />);
  });

  it('should mount and render correctly', () => {
    expect(reportHelpers.parseSearch).toHaveBeenCalledWith(testProps.location.search);
    expect(testProps.addFilters).toHaveBeenCalledWith(filters);
    expect(testProps.refreshReportOptions).toHaveBeenCalledWith(options);
    expect(testProps.initTypeaheadCache).toHaveBeenCalledTimes(1);
    expect(wrapper).toMatchSnapshot();
  });

  describe('with active filters', () => {
    beforeEach(() => {
      wrapper.setProps({
        reportOptions: {
          ...testProps.reportOptions,
          filters: [
            { type: 'A', value: 'aaa' },
            { type: 'B', value: 'bbb' },
            { type: 'C', value: 'ccc' },
          ],
        },
      });
    });

    it('should render', () => {
      const tags = wrapper.find(Tag);
      expect(tags).toHaveLength(3);
      expect(wrapper).toMatchSnapshot();
    });

    it('should remove a filter by index', () => {
      const tags = wrapper.find(Tag);
      tags.first().simulate('remove');
      expect(testProps.removeFilter).toHaveBeenCalledWith(0);
    });
  });

  it('should select a typeahead item', () => {
    const typeahead = wrapper.find(Typeahead);
    const item = {};
    typeahead.simulate('select', item);
    expect(testProps.addFilters).toHaveBeenCalledWith([item]);
  });

  it('should render custom reports when the option is enabled', () => {
    wrapper.setProps({ customReportsEnabled: true });
    expect(wrapper.find('withRouter(Connect(CustomReports))')).toExist();
  });

  it('should mount and render Timezone and Precision Selectors', () => {
    expect(wrapper.find('TimezoneTypeahead')).toExist();
    expect(wrapper.find('PrecisionSelector')).toExist();
  });

  it('should mount and render metrics correctly with only the display selector', () => {
    wrapper.setState({ shownPrecision: 'hour' });
    wrapper.update();
    expect(wrapper.find('PrecisionSelector')).not.toExist();
    expect(wrapper.find('Select')).toExist();
    expect(wrapper.find('Select')).toHaveValue('hour');
  });

  it('should set the disabledAndUTCOnly prop when shownPrecision is day, month, or week', () => {
    wrapper.setState({ shownPrecision: 'day' });
    wrapper.update();

    expect(wrapper.find('TimezoneTypeahead')).toHaveProp('disabledAndUTCOnly', true);

    wrapper.setState({ shownPrecision: 'week' });
    wrapper.update();
    expect(wrapper.find('TimezoneTypeahead')).toHaveProp('disabledAndUTCOnly', true);

    wrapper.setState({ shownPrecision: 'month' });
    wrapper.update();
    expect(wrapper.find('TimezoneTypeahead')).toHaveProp('disabledAndUTCOnly', true);

    wrapper.setState({ shownPrecision: '' });
    wrapper.update();
    expect(wrapper.find('TimezoneTypeahead')).toHaveProp('disabledAndUTCOnly', false);
  });
});
