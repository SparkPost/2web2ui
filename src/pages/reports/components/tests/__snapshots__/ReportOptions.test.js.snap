// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Report Options should mount and render correctly 1`] = `
<div
  data-id="report-options"
>
  <Panel.LEGACY>
    <Panel.LEGACY.Section>
      <Grid>
        <Grid.Column
          md={6}
          xs={9}
        >
          <Connect(Typeahead)
            items={Array []}
            onSelect={[Function]}
            placeholder="Filter by domain, campaign, etc"
            reportOptions={
              Object {
                "filters": Array [],
                "from": 2018-02-15T12:00:00.000Z,
                "relativeRange": "day",
                "to": 2018-02-15T12:00:00.000Z,
              }
            }
            selected={Array []}
          />
        </Grid.Column>
        <Grid.Column
          md={2}
          mdOffset={4}
          xs={3}
        >
          <withRouter(Connect(ShareModal)) />
        </Grid.Column>
      </Grid>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <Grid>
        <Grid.Column
          md={6}
          xs={12}
        >
          <div
            className="FieldWrapperMetricsRollup"
          >
            <DatePicker
              filters={Array []}
              from={2018-02-15T12:00:00.000Z}
              id="date-picker"
              label="Date Range"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {},
                    ],
                  ],
                  "results": undefined,
                }
              }
              preventFuture={true}
              relativeDateOptions={
                Array [
                  "hour",
                  "day",
                  "7days",
                  "30days",
                  "90days",
                  "custom",
                ]
              }
              relativeRange="day"
              roundToPrecision={true}
              selectPrecision={true}
              to={2018-02-15T12:00:00.000Z}
              updateShownPrecision={[Function]}
            />
          </div>
        </Grid.Column>
        <Grid.Column
          md={4}
          xs={6}
        >
          <div
            className="TimezoneTooltipWrapper"
          >
            <Tooltip
              content="Day, week, and month precision only support UTC."
              disabled={true}
            >
              <TimezoneTypeahead
                disabledAndUTCOnly={false}
                onChange={[Function]}
              />
            </Tooltip>
          </div>
        </Grid.Column>
        <Grid.Column
          md={2}
          xs={6}
        >
          <PrecisionSelector
            changeTime={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {},
                  ],
                ],
                "results": undefined,
              }
            }
            from={2018-02-15T12:00:00.000Z}
            to={2018-02-15T12:00:00.000Z}
          />
        </Grid.Column>
      </Grid>
    </Panel.LEGACY.Section>
  </Panel.LEGACY>
</div>
`;

exports[`Component: Report Options with active filters should render 1`] = `
<div
  data-id="report-options"
>
  <Panel.LEGACY>
    <Panel.LEGACY.Section>
      <Grid>
        <Grid.Column
          md={6}
          xs={9}
        >
          <Connect(Typeahead)
            items={Array []}
            onSelect={[Function]}
            placeholder="Filter by domain, campaign, etc"
            reportOptions={
              Object {
                "filters": Array [
                  Object {
                    "type": "A",
                    "value": "aaa",
                  },
                  Object {
                    "type": "B",
                    "value": "bbb",
                  },
                  Object {
                    "type": "C",
                    "value": "ccc",
                  },
                ],
                "from": 2018-02-15T12:00:00.000Z,
                "relativeRange": "day",
                "to": 2018-02-15T12:00:00.000Z,
              }
            }
            selected={
              Array [
                Object {
                  "type": "A",
                  "value": "aaa",
                },
                Object {
                  "type": "B",
                  "value": "bbb",
                },
                Object {
                  "type": "C",
                  "value": "ccc",
                },
              ]
            }
          />
        </Grid.Column>
        <Grid.Column
          md={2}
          mdOffset={4}
          xs={3}
        >
          <withRouter(Connect(ShareModal)) />
        </Grid.Column>
      </Grid>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <Grid>
        <Grid.Column
          md={6}
          xs={12}
        >
          <div
            className="FieldWrapperMetricsRollup"
          >
            <DatePicker
              filters={
                Array [
                  Object {
                    "type": "A",
                    "value": "aaa",
                  },
                  Object {
                    "type": "B",
                    "value": "bbb",
                  },
                  Object {
                    "type": "C",
                    "value": "ccc",
                  },
                ]
              }
              from={2018-02-15T12:00:00.000Z}
              id="date-picker"
              label="Date Range"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {},
                    ],
                  ],
                  "results": undefined,
                }
              }
              preventFuture={true}
              relativeDateOptions={
                Array [
                  "hour",
                  "day",
                  "7days",
                  "30days",
                  "90days",
                  "custom",
                ]
              }
              relativeRange="day"
              roundToPrecision={true}
              selectPrecision={true}
              to={2018-02-15T12:00:00.000Z}
              updateShownPrecision={[Function]}
            />
          </div>
        </Grid.Column>
        <Grid.Column
          md={4}
          xs={6}
        >
          <div
            className="TimezoneTooltipWrapper"
          >
            <Tooltip
              content="Day, week, and month precision only support UTC."
              disabled={true}
            >
              <TimezoneTypeahead
                disabledAndUTCOnly={false}
                onChange={[Function]}
              />
            </Tooltip>
          </div>
        </Grid.Column>
        <Grid.Column
          md={2}
          xs={6}
        >
          <PrecisionSelector
            changeTime={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {},
                  ],
                ],
                "results": undefined,
              }
            }
            from={2018-02-15T12:00:00.000Z}
            to={2018-02-15T12:00:00.000Z}
          />
        </Grid.Column>
      </Grid>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <small>
        Filters:
      </small>
      <Tag
        className="TagWrapper"
        key="0"
        onRemove={[Function]}
      >
        A
        : 
        aaa
      </Tag>
      <Tag
        className="TagWrapper"
        key="1"
        onRemove={[Function]}
      >
        B
        : 
        bbb
      </Tag>
      <Tag
        className="TagWrapper"
        key="2"
        onRemove={[Function]}
      >
        C
        : 
        ccc
      </Tag>
    </Panel.LEGACY.Section>
  </Panel.LEGACY>
</div>
`;
